programmatically create string content (#16908)
dotnet/docs: docs/csharp/how-to/modify-string-contents.md

[!code-csharp-interactive[replace creates a new string](../../../samples/snippets/csharp/how-to/strings/ModifyStrings.cs#6)]

> ## Programmatically building up string content

Reviewer: ```suggestion
## Programmatically build up string content
```

> Since strings are immutable, the above methods all create temporary strings or character arrays. In high performance scenarios it may be desirable to avoid these heap allocations. .NET core added a <xref:System.String.Create> method that allows you to programmatically fill in the character content of a string via a callback while avoiding the intermediate temporary string allocations. 

Reviewer: ```suggestion
Since strings are immutable, the previous examples all create temporary strings or character arrays. In high performance scenarios, it may be desirable to avoid these heap allocations. .NET core added a <xref:System.String.Create> method that allows you to programmatically fill in the character content of a string via a callback while avoiding the intermediate temporary string allocations. 
```

Do we have a code sample that uses string.Create?

> After a string is created, you could use **unsafe** code to modify its content but this is strongly discouraged. Unsafe code bypasses many of the features of .NET designed to minimize certain types of bugs in code and may prevent using some features that only work with normal code. The following example is provided for those rare situations where you want to modify a string in-place using unsafe code. The example shows how to use the `fixed` keyword. The `fixed` keyword prevents the garbage collector (GC) from moving the string object in memory while code accesses the memory using the unsafe pointer. It also demonstrates one possible side effect of unsafe operations on strings that results from the way that the C# compiler stores (interns) strings internally. In general, you shouldn't use this technique unless it is absolutely necessary. You can learn more in the articles on [unsafe](../language-reference/keywords/unsafe.md) and [fixed](../language-reference/keywords/fixed-statement.md). The API reference for <xref:System.String.Intern%2A> includes information on string interning.

Reviewer: ```suggestion
After a string is created, you could use **unsafe** code to modify its content, but this is strongly discouraged. Unsafe code bypasses many of the features of .NET designed to minimize certain types of bugs in code and may prevent using some features that only work with normal code. The following example is provided for those rare situations where you want to modify a string in-place using unsafe code. The example shows how to use the `fixed` keyword. The `fixed` keyword prevents the garbage collector from moving the string object in memory while code accesses the memory using the unsafe pointer. It also demonstrates one possible side effect of unsafe operations on strings that results from the way that the C# compiler stores (interns) strings internally. In general, you shouldn't use this technique unless it is absolutely necessary. You can learn more in the articles on [unsafe](../language-reference/keywords/unsafe.md) and [fixed](../language-reference/keywords/fixed-statement.md). The API reference for <xref:System.String.Intern%2A> includes information on string interning.
```

Reviewer: I removed the GC acronym since it's not mentioned after this at all, so it's unnecessary.

[!code-csharp-interactive[replace creates a new string](../../../samples/snippets/csharp/how-to/strings/ModifyStrings.cs#6)]

## Programmatically build up string content

> Since strings are immutable, the previous examples all create temporary strings or character arrays. In high performance scenarios, it may be desirable to avoid these heap allocations. .NET core added a <xref:System.String.Create> method that allows you to programmatically fill in the character content of a string via a callback while avoiding the intermediate temporary string allocations.

Reviewer 2: ```suggestion
Since strings are immutable, the previous examples all create temporary strings or character arrays. In high performance scenarios, it may be desirable to avoid these heap allocations. .NET Core added a <xref:System.String.Create%2A?displayProperty=nameWithType> method that allows you to programmatically fill in the character content of a string via a callback while avoiding the intermediate temporary string allocations.
```

> Since strings are immutable, the previous examples all create temporary strings or character arrays. In high performance scenarios, it may be desirable to avoid these heap allocations. .NET Core added a <xref:System.String.Create%2A?displayProperty=nameWithType> method that allows you to programmatically fill in the character content of a string via a callback while avoiding the intermediate temporary string allocations.

Reviewer: ```suggestion
Since strings are immutable, the previous examples all create temporary strings or character arrays. In high performance scenarios, it may be desirable to avoid these heap allocations. .NET Core provides a <xref:System.String.Create%2A?displayProperty=nameWithType> method that allows you to programmatically fill in the character content of a string via a callback while avoiding intermediate temporary string allocations.
```

You could modify a string in a fixed block with unsafe code but it is **strongly** discouraged to modify the string content after a string is created. Doing so will break things in unpredictable ways. For example, if someone interns a string that has the same content as yours, they will get your copy and will not expect that you are modifying their string at all.

> [!code-csharp[using string.Create to programmatically build the string content for a new string](../../../samples/snippets/csharp/how-to/strings/ModifyStrings.cs#7)]

Reviewer 2: There is unrelated paragraph between the example and the paragraph that describes the example. Should the example be moved one paragraph up?

Author: sounds good; I'll move it up
